ENH: Close memmap file with a public function close()

yoonghm opened the pull request:
Create a public function close() to properly close mmap

See Method to explicitly close memmap #13510, there are two ways to un-mapped a memory-mapped file:
a. del fp or
b. fp._mmap.close()

The first method may leave the file locked for a long period of time. The second method exposes internal variable _mmap.

The solution provides public function to close the memmap object.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

yoonghm committed:
ENH: Close memmap file with a public function close()

Create a public function close() to properly close mmap

See Method to explicitly close memmap #13510, there are two ways to un-mapped a memory-mapped file:
a. del fp or
b. fp._mmap.close()

The first method may leave the file locked for a long period of time. The second method exposes internal variable _mmap.

The solution provides public function to close the memmap object.

yoonghm commented:
I have tested it again on Windows 10, it did not crash when I call `fp.close()`.

BvB93 commented:
Please don't recreate the pull request every time you make an addition/modification,
just push the respective change.

eric-wieser commented:
Xref #17940, note the comment about crashing the interpreter there may still apply

BvB93 commented on numpy/core/memmap.py:
```diff
@@ -335,3 +341,21 @@ def __getitem__(self, index):
+        
+        It has no effect if no file is memory-mapped.
+        For further information, see `memmap`.
+        Parameters
```
```suggestion
        
        Parameters
```
An empty line is missing here; same for the "See Also" section.

charris commented:
This will need a release note.

yoonghm committed:
Update numpy/core/memmap.py

Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

BvB93 commented on numpy/core/memmap.py:
```diff
@@ -335,3 +341,22 @@ def __getitem__(self, index):
+        Parameters
+        ----------
+        None
+        See Also
```
```suggestion

        See Also
```
As was mentioned previously, an empty line is missing here as well.

yoonghm committed:
Update numpy/core/memmap.py

Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

yoonghm commented:
I am not familiar with the github GUI. That's why re-submit similar pull requests. Thanks

yoonghm committed:
Added context manager for memmap and closed()

Added context manager and close() function.  I have created an example to test this features.

BvB93 commented on numpy/core/memmap.py:
```diff
@@ -335,3 +341,48 @@ def __getitem__(self, index):
+        memmap
+
+        """
+        return True if self._mmap is None else False
```
```suggestion
        return True if self._mmap is None else self._mmap.closed
```
If `self._mmap is not None` it does not necessarily mean that the `mmap.mmap` is still opened.

yoonghm replied:
You are right. Thanks.

yoonghm committed:
Update memmap.py

Update example using context manager and closed()

yoonghm committed:
Update numpy/core/memmap.py

Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

BvB93 commented on numpy/core/memmap.py:
```diff
@@ -335,3 +361,48 @@ def __getitem__(self, index):
+        """
+        if self._mmap is not None:
+            self._mmap.close()
+            self._mmap = None
```
Is there a reason for reassigning `_mmap` as `None`?

yoonghm committed:
Update memmap.py

BvB93 commented on numpy/core/memmap.py:
```diff
@@ -201,6 +201,32 @@ class memmap(ndarray):
+    Use context manager to read the file into memmap:
+    
+    >>> with np.memmap(filename, dtype='float32', mode='r') as fpcon:
+    >>>     fpcon
```
```suggestion
    ...     fpcon
```

BvB93 replied:
This should fix the failing doctest.

eric-wieser reviewed (changes requested):
As discussed in #13510, this is not safe, and calling `close()` on a `memmap` for which a view survives elsewhere causes a segfault.

yoonghm committed:
Update numpy/core/memmap.py

Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

yoonghm commented:
> As discussed in #13510, this is not safe, and calling `close()` on a `memmap` for which a view survives elsewhere causes a segfault.

I have tested it, the code work fine.

```Python
import os.path
import shutil
from tempfile import mkdtemp
import numpy as np


np.random.seed(1234)
data = np.random.uniform(0, 100, 4)
data.resize(4, 1)

tempdir = mkdtemp()
filename = os.path.join(tempdir, 'newfile.dat')
print(f'{filename = }\n')

fp1 = np.memmap(filename, dtype='float64', mode='w+', shape=(4,1))
fp1[:] = data[:]
fp1.flush()
print(f'{fp1 = }\n')

fp2 = np.memmap(filename, dtype='float64', mode='r+', shape=(4,1))
print(f'{fp2 = }\n')

# Close fp1 but fp2 is still pointing to the file
fp1.close()

# Perform write to fp2
fp2[:] = data*data
fp2.flush()
print(f'{fp2 = }\n')

# See what is inside filename using context manager
with np.memmap(filename, dtype='float64', mode='r') as fp3:
    print(f'{fp3 = }\n')

# Close fp2 now
fp2.close()

# Delete everything
shutil.rmtree(tempdir)
```

The output is

```
filename = 'C:\\Users\\yoonghm\\AppData\\Local\\Temp\\tmpdgsiz4pb\\newfile.dat'

fp1 = memmap([[19.15194504],
        [62.2108771 ],
        [43.7727739 ],
        [78.53585837]])

fp2 = memmap([[19.15194504],
        [62.2108771 ],
        [43.7727739 ],
        [78.53585837]])

fp2 = memmap([[ 366.79699873],
        [3870.19323005],
        [1916.05573496],
        [6167.88105013]])

fp3 = memmap([ 366.79699873, 3870.19323005, 1916.05573496, 6167.88105013])
```

eric-wieser commented:
> I have tested it, the code work fine.

That is because you're not running the right test. I expect the one in https://github.com/numpy/numpy/issues/13510#issuecomment-741972746 to still fail (after replacing `_mmap.close` with `close`)

yoonghm committed:
Update memmap.py

Update output from with context manager for unit testing.

yoonghm committed:
Update memmap.py

yoonghm commented:
I have fixed the unit test within the code. However, do not know how to test the code again.

```Python
import os.path
import shutil
from tempfile import mkdtemp
import numpy as np


np.random.seed(1234)
data = np.random.uniform(0, 100, 4)
data

tempdir = mkdtemp()
filename = os.path.join(tempdir, 'newfile.dat')
print(f'{filename = }\n')

fp1 = np.memmap(filename, dtype='float64', mode='w+', shape=(4,1))
fp1[:] = data[:]
fp1.flush()
print(f'{fp1 = }\n')

fp2 = np.memmap(filename, dtype='float64', mode='r+', shape=(4,1))
print(f'{fp2 = }\n')

# Close fp1 but fp2 is still pointing to the file
fp1.close()

# Perform write to fp2
fp2[:] = data*data
fp2.flush()
print(f'{fp2 = }\n')

# See what is inside filename using context manager
with np.memmap(filename, dtype='float64', mode='r') as fp3:
    print(f'{fp3 = }\n')

# Close fp2 now
fp2.close()

# Delete everything
shutil.rmtree(tempdir)
```

I have tested the code. Do not know why my last comment was not shown up here.

yoonghm commented:
See my comment in #13510
